utils/utils.c

```diff
@@ -392,3 +392,81 @@ char * strjoin(char *left, char *right, char *delim)
 	strcpy(new + len - rlen - 1, right);
 	return new;
 }
+
+#define QUOTE '\''
+#define DQUOTE '"'
+
+static int setargs(char *args, char **argv)
+{
+	int count = 0;
+
+	/* ignore start spaces */
+	while (isspace(*args))
+		++args;
+
+	while (*args) {
+		/* consider quotes and update argv */
+		if (*args == QUOTE) {
+			if (argv)
+				argv[count] = args + 1;
+			while (*args != QUOTE)
+				++args;
+		} else if (*args == DQUOTE) {
+			if (argv)
+				argv[count] = args + 1;
+			while (*args != DQUOTE)
+				++args;
+		} else if (argv) {
+			argv[count] = args;
+		}
+		/* read characters until '\0' or space */
+		while (*args && !isspace(*args))
+			++args;
+		/* set '\0' rather than space */
+		if (argv && *args)
+			*args++ = '\0';
+		/* ignore margin spaces */
+		while (isspace(*args))
+			++args;
+		/* count up argument */
+		count++;
+	}
+
+	if (argv)
+		argv[count] = NULL;
+	return count;
+}
+
+#undef QUOTE
+#undef DQUOTE
+
+char **parse_cmdline(char *cmd, int *argc)
+{
+	char **argv = NULL;
+	int argn = 0;
+
+	if (cmd && *cmd
+		&& (cmd = strdup(cmd))
+		&& (argn = setargs(cmd, NULL))
+		&& (argv = malloc((argn + 1) * sizeof(char *))))
+	{
+		*argv++ = cmd;
+		argn = setargs(cmd, argv);
+	}
+
+	if (cmd && !argv)
+		free(cmd);
+
+	if (argc)
+		*argc = argn;
+
+	return argv;
+}
+
+void free_parsed_args(char **argv)
+{
+	if (argv) {
+		free(argv[-1]);
+		free(argv - 1);
+	}
```

Comment:
This doesn't look straight-forward.  You could either pre-decrement the "argv" with a comment or make `parse_cmdline()` return "&argv[0]" and use "argv[1...]" in `execute_recv_cmd()`.

 Also I think it'd be better making the function name consistent - like `free_parsed_cmdline()`?